doc.go

```diff
@@ -27,13 +27,335 @@
 // to use struct tags or embed special types, so Fx automatically works well
 // with most Go packages.
 //
-// Basic usage is explained in the package-level example below. If you're new
-// to Fx, start there! Advanced features, including named instances, optional
-// parameters, and value groups, are explained under the In and Out types.
+// # Basic usage
+//
+// Basic usage is explained in the package-level example.
+// If you're new to Fx, start there!
+//
+// Advanced features, including named instances, optional parameters,
+// and value groups, are explained in this section further down.
 //
 // # Testing Fx Applications
 //
 // To test functions that use the Lifecycle type or to write end-to-end tests
 // of your Fx application, use the helper functions and types provided by the
 // go.uber.org/fx/fxtest package.
+//
+// # Parameter Structs
+//
+// Fx constructors declare their dependencies as function parameters. This can
+// quickly become unreadable if the constructor has a lot of dependencies.
+//
+//	func NewHandler(users *UserGateway, comments *CommentGateway, posts *PostGateway, votes *VoteGateway, authz *AuthZGateway) *Handler {
+//		// ...
+//	}
+//
+// To improve the readability of constructors like this, create a struct that
+// lists all the dependencies as fields and change the function to accept that
+// struct instead. The new struct is called a parameter struct.
+//
+// Fx has first class support for parameter structs: any struct embedding
+// fx.In gets treated as a parameter struct, so the individual fields in the
+// struct are supplied via dependency injection. Using a parameter struct, we
+// can make the constructor above much more readable:
+//
+//	type HandlerParams struct {
+//		fx.In
+//
+//		Users    *UserGateway
+//		Comments *CommentGateway
+//		Posts    *PostGateway
+//		Votes    *VoteGateway
+//		AuthZ    *AuthZGateway
+//	}
+//
+//	func NewHandler(p HandlerParams) *Handler {
+//		// ...
+//	}
+//
+// Though it's rarely a good idea, constructors can receive any combination of
```

Comment:
Should we maybe add a bit more detail here on why this is not a good idea?

Reply:
I guess it's less of a "good idea" thing and more of a "unnecessary."?
I'll change the wording here.